File: schemas/analytics-metrics/convex/experiment_results.ts
// experiment_results: Statistical results per variant.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const experiment_results = defineTable({
  experimentId: v.id("experiments"),
  variantId: v.id("experiment_variants"),
  metricId: v.id("metric_definitions"),
  sampleSize: v.number(),
  meanValue: v.optional(v.number()),
  stddev: v.optional(v.number()),
  ciLower: v.optional(v.number()),
  ciUpper: v.optional(v.number()),
  pValue: v.optional(v.number()),
  lift: v.optional(v.number()),
  isSignificant: v.boolean(),
  computedAt: v.number(),
  updatedAt: v.number(),
})
  .index("by_experiment_id_and_variant_id_and_metric_id", ["experimentId", "variantId", "metricId"]);

File: schemas/analytics-metrics/spacetimedb/experiment_results.rs
// experiment_results: Statistical results per variant per metric with significance testing.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

// Composite unique constraint (not expressible inline):
// - unique(experiment_id, variant_id, metric_id)

#[spacetimedb::table(name = experiment_results, public)]
pub struct ExperimentResult {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub experiment_id: String, // UUID, FK → experiments.id (cascade delete)

    pub variant_id: String, // UUID, FK → experiment_variants.id (cascade delete)
    pub metric_id: String, // UUID, FK → metric_definitions.id (cascade delete)
    pub sample_size: i64,
    pub mean_value: Option<f64>,
    pub stddev: Option<f64>,
    pub ci_lower: Option<f64>,
    pub ci_upper: Option<f64>,
    pub p_value: Option<f64>,
    pub lift: Option<f64>,
    pub is_significant: bool,
    pub computed_at: Timestamp,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
